crates/worker/src/partition_processor_manager.rs

```diff
@@ -743,6 +749,45 @@ impl<T: TransportConnect> PartitionProcessorManager<T> {
         }
     }
 
+    fn request_partition_snapshots(&mut self) {
+        let Some(records_per_snapshot) = self
+            .updateable_config
+            .live_load()
+            .worker
+            .snapshots
+            .records_per_snapshot
+        else {
+            return;
+        };
+
+        for (partition_id, status) in self.running_partition_processors.iter() {
+            let ProcessorStatus::Started(state) = status else {
+                continue;
+            };
+            let status = state.watch_rx.borrow().clone();
+            if status.is_effective_leader()
+                && status.replay_status == ReplayStatus::Active
+                && status.last_applied_log_lsn.unwrap_or(Lsn::INVALID)
+                    >= status
+                        .last_archived_log_lsn
+                        .unwrap_or(Lsn::OLDEST)
+                        .add(Lsn::from(records_per_snapshot))
+            {
+                debug!(%partition_id, "Triggering snapshot on partition leader after records_per_snapshot exceeded (last snapshot: {}, current LSN: {})", status.last_archived_log_lsn.unwrap_or(Lsn::OLDEST), status.last_applied_log_lsn.unwrap_or(Lsn::INVALID));
+                let (tx, _) = oneshot::channel();
+
+                // ignore errors and don't request further snapshots if internal queue is full; we will try again later
+                if self
+                    .tx
+                    .try_send(ProcessorsManagerCommand::CreateSnapshot(*partition_id, tx))
+                    .is_err()
+                {
+                    break;
+                }
```

Comment:
No strong reason to do so immediately, though two weak intuitions lead me to think that `ParitionProcessor` is the better owner of this responsibility than the `PartitionProcessorManager`:

- PP is the logical owner of the partition's `PartitionStore` once running. Sure, we can clone the handle and pass references around, but I feel that I want to keep interactions with it mostly clustered in the `PartitionProcessor`. To my eyes this is different to restoring a snapshot on bootstrap, which happens before we start the `PartitionProcessor` (hence why that responsibility currently resides in `PartitionStoreManager::restore_partition_store_snapshot`)
- at some point we might want to create snapshots from a deterministic LSN by sending a `CreateSnapshot` command through the log, and `PartitionProcessor` is the place to handle that

To be sure, neither of these is a very strong argument - and I'm not attached to it remaining here! It's a pretty easy change to make given that the snapshot logic mostly lives in `SnapshotProducer`.

Reply:
Regarding creating snapshots for deterministic LSNs, I think we need a few more things (a) being able to create a snapshot for a given LSN and (b) probably the globally total ordered log to be able to give a LSN a global meaning. Given that it's not very clear when these things are going to happen, I wouldn't over index on it too much right now. Moreover, it could also be solved differently by a combination of snapshots and a log segment that brings all nodes to a specific LSN.